### PR TITLE
Expand, rename and improve `incorrect_fn_null_checks` lint

### LINKED ISSUE #113601: Extend incorrect_fn_null_checks lint to useless_ptr_null_checks
I'd suggest extending the newly added `incorrect_fn_null_checks` lint added in #111717 to cover cases of null comparison for other non-nullable types, like for instance in:

```Rust
    let v: &u8 = &0;
    (v as *const u8).is_null();
```

This would involve a renaming of the lint to `useless_ptr_null_checks` or `incorrect_ptr_null_checks`. I don't think having multiple lints for different kinds of non-nullable types is useful (one for function pointers, one for references, etc).

See discussion [starting from here](https://github.com/rust-lang/rust/pull/111717#issuecomment-1565341476).

cc @Urgau @scottmcm

### PR DESCRIPTION
**Pull request by Urgau**
This PR,

 - firstly, expand the lint by now linting on references
 - secondly, it renames the lint `incorrect_fn_null_checks` -> `useless_ptr_null_checks`
 - and thirdly it improves the lint by catching `ptr::from_mut`, `ptr::from_ref`, as well as `<*mut _>::cast` and `<*const _>::cast_mut`

Fixes https://github.com/rust-lang/rust/issues/113601
cc @est31

### CODE REVIEW

**Comment by est31** [comment]
Very good work, in the future it would be also good to lint on these:

* `Box::into_raw`
* `NonNull::as_ptr`
* `str::as_ptr`/`str::as_mut_ptr`
* `Vec::as_ptr`/`Vec::as_mut_ptr`
* `slice::as_ptr`/`slice::as_mut_ptr`
* `{Cell,Weak,Rc,Arc}::as_ptr`
* `CStr::as_ptr`

etc. Maybe it might make sense to introduce a `#[rustc_pointer_non_null]` attribute and add it to these functions?

That's more for follow up work though.

When working on #113652 I was also briefly considering using `ty_is_known_nonnull` from types.rs, but that one is about types and what we really need to check for is conversions, whether it's a cast or one of the known functions. We can't just take any function on these types with a pointer return value, the function might actually do something completely different. In other words, the approach of this PR is good.

**Comment by est31** [comment]
> That's more for follow up work though.

Especially I'd like this to get merged sooner than later as it would be nice if the lint rename wasn't exposed to stable users. Beta has just branched, so we have weeks of time, but still.

File: compiler/rustc_lint/src/ptr_nulls.rs [hunk 1]
```diff
@@ -22,27 +22,56 @@ declare_lint! {
...
     ///
-    /// Function pointers are assumed to be non-null, checking them for null will always
-    /// return false.
-    INCORRECT_FN_NULL_CHECKS,
+    /// Function pointers and references are assumed to be non-null, checking them for null
+    /// will always return false.
+    USELESS_PTR_NULL_CHECKS,
     Warn,
-    "incorrect checking of null function pointer"
+    "useless checking of always null pointer"
```
Comment (est31) [inline compiler/rustc_lint/src/ptr_nulls.rs:29]:
```suggestion
    "useless checking of non-null-typed pointer"
```

**Review by est31 — APPROVED** [review]

**Comment by est31** [comment]
Approved this but won't r+ as I think it would be better for someone from the compiler team to approve it -- note that [as per @scottmcm ](https://github.com/rust-lang/rust/pull/111717#issuecomment-1565665644) this doesn't need a lang team FCP.

File: compiler/rustc_lint/src/ptr_nulls.rs [hunk 2]
```diff
@@ -22,27 +22,56 @@ declare_lint! {
...
-    "incorrect checking of null function pointer"
+    "useless checking of non-null-typed pointer"
 }
 
-declare_lint_pass!(IncorrectFnNullChecks => [INCORRECT_FN_NULL_CHECKS]);
+declare_lint_pass!(PtrNullChecks => [USELESS_PTR_NULL_CHECKS]);
 
-fn is_fn_ptr_cast(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
-    let mut expr = expr.peel_blocks();
+fn ptr_cast_chain<'a>(cx: &'a LateContext<'_>, mut e: &'a Expr<'a>) -> Option<&'a Expr<'a>> {
```
Comment (cjgillot) [inline compiler/rustc_lint/src/ptr_nulls.rs:34]: Could you document what this method computes and what it returns? That's not obvious just looking at it.

File: compiler/rustc_lint/src/ptr_nulls.rs [hunk 3]
```diff
@@ -22,27 +22,56 @@ declare_lint! {
...
-    while let ExprKind::Cast(cast_expr, cast_ty) = expr.kind
-            && let TyKind::Ptr(_) = cast_ty.kind {
-        expr = cast_expr.peel_blocks();
-        had_at_least_one_cast = true;
+    loop {
+        e = e.peel_blocks();
+        e = if let ExprKind::Cast(e, t) = e.kind
+            && let TyKind::Ptr(_) = t.kind {
+            had_at_least_one_cast = true;
+            e
```
Comment (cjgillot) [inline compiler/rustc_lint/src/ptr_nulls.rs:41]: Nit: could you rename the pattern here, it looks like `e = e`.

File: compiler/rustc_lint/src/ptr_nulls.rs [hunk 4]
```diff
@@ -22,27 +22,56 @@ declare_lint! {
...
+        } else {
+            return None;
+        };
     }
-    had_at_least_one_cast && cx.typeck_results().expr_ty_adjusted(expr).is_fn()
 }
 
-impl<'tcx> LateLintPass<'tcx> for IncorrectFnNullChecks {
+fn incorrect_check<'a>(cx: &LateContext<'a>, expr: &Expr<'_>) -> Option<PtrNullChecksDiag<'a>> {
+    ptr_cast_chain(cx, expr).and_then(|expr| {
```
Comment (cjgillot) [inline compiler/rustc_lint/src/ptr_nulls.rs:62]: Nit: could you prefer `?`

File: compiler/rustc_lint/src/lints.rs [hunk 5]
```diff
@@ -613,11 +613,19 @@ pub struct ExpectationNote {
...
-// fn_null_check.rs
+// ptr_nulls.rs
 #[derive(LintDiagnostic)]
-#[diag(lint_fn_null_check)]
-#[help]
-pub struct FnNullCheckDiag;
+pub enum PtrNullChecksDiag<'a> {
+    #[diag(lint_ptr_null_checks_fn_ptr)]
+    #[help(lint_help)]
+    FnPtr,
```
Comment (cjgillot) [inline compiler/rustc_lint/src/lints.rs:621]: Should the expression's span and type be added here too, for symmetry?

File: compiler/rustc_lint/src/ptr_nulls.rs (hunk 2 above)
Comment (Urgau) [reply compiler/rustc_lint/src/ptr_nulls.rs:36]: Done.

File: compiler/rustc_lint/src/lints.rs (hunk 5 above)
Comment (Urgau) [reply compiler/rustc_lint/src/lints.rs]: Don't know, but added anyway since it doesn't make the output any worse.

**Pull request merged**